### PR TITLE
feat: add constraint filters

### PR DESCRIPTION
**Pull request by lohart13**
## Pull Request Checklist

- [x] Have you added new tests to prevent regressions?
- [ ] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description Of Change

This PR adds constraint filters to the `showConstraint` methods and fixes some issues with the `showConstraints` methods where data wasn't being returned or was been parsed incorrectly.

### CODE REVIEW

**Comment by lohart13** [comment]
No idea why the mssql oldest test has failed.
This passes fine in my local dev environment.

**Comment by WikiRik** [comment]
> No idea why the mssql oldest test has failed. This passes fine in my local dev environment.

That one is a flaky test, see #14161

File: packages/core/src/dialects/db2/query-generator-typescript.ts [hunk 1]
```diff
@@ -29,32 +30,49 @@ export class Db2QueryGeneratorTypeScript extends AbstractQueryGenerator {
...
+        throw new Error(`Constraint type ${type} is not supported`);
+    }
+  }
+
   showConstraintsQuery(tableName: TableNameOrModel, options?: ShowConstraintsQueryOptions) {
     const table = this.extractTableDetails(tableName);
 
     return joinSQLFragments([
-      'SELECT c.TABSCHEMA AS "constraintSchema",',
+      'SELECT TRIM(c.TABSCHEMA) AS "constraintSchema",',
```
Comment (WikiRik) [inline packages/core/src/dialects/db2/query-generator-typescript.ts:52]: Any specific reason why you are adding `TRIM()` to the schemas and not the other values?

File: packages/core/src/dialects/db2/query-generator-typescript.ts [hunk 2]
```diff
@@ -29,32 +30,49 @@ export class Db2QueryGeneratorTypeScript extends AbstractQueryGenerator {
...
+      'LEFT JOIN SYSCAT.KEYCOLUSE k ON c.CONSTNAME = k.CONSTNAME AND c.TABNAME = k.TABNAME AND c.TABSCHEMA = k.TABSCHEMA',
       'LEFT JOIN SYSCAT.KEYCOLUSE fk ON r.REFKEYNAME = fk.CONSTNAME',
       'LEFT JOIN SYSCAT.CHECKS ck ON c.CONSTNAME = ck.CONSTNAME AND c.TABNAME = ck.TABNAME AND c.TABSCHEMA = ck.TABSCHEMA',
       `WHERE c.TABNAME = ${this.escape(table.tableName)}`,
       'AND c.TABSCHEMA =',
       table.schema ? this.escape(table.schema) : 'USER',
+      options?.columnName ? `AND k.COLNAME = ${this.escape(options.columnName)}` : '',
       options?.constraintName ? `AND c.CONSTNAME = ${this.escape(options.constraintName)}` : '',
-      'ORDER BY c.CONSTNAME',
+      options?.constraintType ? `AND c.TYPE = ${this.escape(this._getConstraintType(options.constraintType))}` : '',
```
Comment (WikiRik) [inline packages/core/src/dialects/db2/query-generator-typescript.ts:74]:
```suggestion
      options?.constraintType ? `AND c.TYPE = ${this._getConstraintType(options.constraintType)}` : '',
```
No need to escape values that we control

File: packages/core/src/dialects/mssql/query-generator-typescript.ts [hunk 3]
```diff
@@ -63,27 +81,34 @@ export class MsSqlQueryGeneratorTypeScript extends AbstractQueryGenerator {
...
       ', OBJECT_NAME(k.[referenced_object_id]) AS referencedTableName, rcol.[name] AS referencedColumnNames, k.[delete_referential_action_desc] AS deleteAction',
       ', k.[update_referential_action_desc] AS updateAction, null AS definition FROM sys.foreign_keys k INNER JOIN sys.foreign_key_columns c ON k.[object_id] = c.constraint_object_id',
       'INNER JOIN sys.columns fcol ON c.parent_column_id = fcol.column_id AND c.parent_object_id = fcol.object_id',
       'INNER JOIN sys.columns rcol ON c.referenced_column_id = rcol.column_id AND c.referenced_object_id = rcol.object_id',
       ') c ON t.object_id = c.constraintTableId',
       `WHERE s.name = ${this.escape(table.schema)} AND t.name = ${this.escape(table.tableName)}`,
+      options?.columnName ? `AND c.columnNames = ${this.escape(options.columnName)}` : '',
       options?.constraintName ? `AND c.constraintName = ${this.escape(options.constraintName)}` : '',
+      options?.constraintType ? `AND c.constraintType = ${this.escape(this._getConstraintType(options.constraintType))}` : '',
```
Comment (WikiRik) [inline packages/core/src/dialects/mssql/query-generator-typescript.ts:111]:
```suggestion
      options?.constraintType ? `AND c.constraintType = ${this._getConstraintType(options.constraintType)}` : '',
```

File: packages/core/src/dialects/snowflake/query-generator-typescript.ts [hunk 4]
```diff
@@ -14,6 +14,10 @@ export class SnowflakeQueryGeneratorTypeScript extends AbstractQueryGenerator {
   showConstraintsQuery(tableName: TableNameOrModel, options?: ShowConstraintsQueryOptions) {
     const table = this.extractTableDetails(tableName);
 
+    if (options?.columnName) {
```
Comment (WikiRik) [inline packages/core/src/dialects/snowflake/query-generator-typescript.ts:17]: I think we can replace this with usage of `rejectInvalidOptions` to make it more unified with the other functions

**Review by WikiRik — COMMENTED** [review]
Small few remarks from the query-generator changes. Will look at the query-interface changes (including QI tests) later

File: packages/core/src/dialects/db2/query-generator-typescript.ts (hunk 1 above)
Comment (lohart13) [reply packages/core/src/dialects/db2/query-generator-typescript.ts:52]: The `TABSCHEMA` value seems to return a trailing blank space hence the reason for adding `TRIM()`

File: packages/core/src/dialects/snowflake/query-generator-typescript.ts (hunk 4 above)
Comment (lohart13) [reply packages/core/src/dialects/snowflake/query-generator-typescript.ts]: Done

File: packages/core/src/dialects/db2/query-generator-typescript.ts (hunk 2 above)
Comment (lohart13) [reply packages/core/src/dialects/db2/query-generator-typescript.ts:74]: The value still needs to be quoted hence the reason for using escape.

File: packages/core/src/dialects/mssql/query-generator-typescript.ts (hunk 3 above)
Comment (lohart13) [reply packages/core/src/dialects/mssql/query-generator-typescript.ts:111]: The value still needs to be quoted hence the reason for using escape.

File: packages/core/src/dialects/db2/query-generator-typescript.ts (hunk 2 above)
Comment (WikiRik) [reply packages/core/src/dialects/db2/query-generator-typescript.ts:74]: You'll need to add the proper quotes to `_getConstraintType` as well but maybe that's not the best for readability if it is quoted 'twice'. So this is fine

File: packages/core/test/integration/query-interface/add-show-remove-constraint.test.ts [hunk 5]
```diff
@@ -7,12 +7,17 @@ const dialect = sequelize.getDialect();
 
 describe('QueryInterface#{add,show,removeConstraint}', () => {
   describe('Without schema', () => {
+    const defaultSchema = dialect === 'db2' ? 'DB2INST1' : sequelize.dialect.getDefaultSchema();
```
Comment (WikiRik) [inline packages/core/test/integration/query-interface/add-show-remove-constraint.test.ts:10]: If this is the actual default schema for DB2 we should add it to `getDefaultSchema` in the index.ts file of DB2

**Review by WikiRik — APPROVED** [review]

File: packages/core/test/integration/query-interface/add-show-remove-constraint.test.ts (hunk 5 above)
Comment (lohart13) [reply packages/core/test/integration/query-interface/add-show-remove-constraint.test.ts:10]: According to https://www.ibm.com/docs/en/db2-for-zos/12?topic=concepts-db2-schemas-schema-qualifiers, the default schema is `the authorization ID of the owner of the plan or package`.

Do you know if `the authoriation ID of the owner` is the same as the username that is used in the connection?
In this case the default schema would be equal to the username of the sequelize connection.

Comment (WikiRik) [reply packages/core/test/integration/query-interface/add-show-remove-constraint.test.ts:10]: Unfortunately I don't. But maybe you can play around which some env variables and see which one is changing the schema?

Comment (lohart13) [reply packages/core/test/integration/query-interface/add-show-remove-constraint.test.ts:10]: Ok so I changed the `DB2INSTANCE` in the `.env_list` file in the `dev/db2` folder to a random value and updated this in the scripts and test config and when logging in to the cli and executing the commands, it appears that the default schema is the username of the user.

Do you want to get this verified by someone that actually uses db2 or just update the `getDefaultSchema` to return the username in the sequelize connection config.

Comment (WikiRik) [reply packages/core/test/integration/query-interface/add-show-remove-constraint.test.ts:10]: Just update it, it seems to correspond with the documentation of db2 as well

Comment (lohart13) [reply packages/core/test/integration/query-interface/add-show-remove-constraint.test.ts]: Done

Comment (WikiRik) [reply packages/core/test/integration/query-interface/add-show-remove-constraint.test.ts]: Actually, thinking about this again; can you move this to a separate PR where you also update the other query generator functions that now use `USER` and `CURRENT SCHEMA`? Then that is more aligned with the other dialects

Comment (lohart13) [reply packages/core/test/integration/query-interface/add-show-remove-constraint.test.ts:10]: Sure, have submitted PR #16307 which implements these changes.

Comment (WikiRik) [inline packages/core/test/integration/query-interface/add-show-remove-constraint.test.ts:10]: We can remove this now, right?

Comment (lohart13) [reply packages/core/test/integration/query-interface/add-show-remove-constraint.test.ts:10]: Good catch. We need the constant but not the conditional check.

**Review by WikiRik — APPROVED** [review]

**Pull request merged**